Doesn't work with vcrpy

There's all sort of stuff being broken with [vcrpy](https://github.com/kevin1024/vcrpy/)'s extensive monkeypatching. I have tried fixing it, this is as far as I got (yes, this is a patch for vcrpy code):

```python
diff --git a/vcr/patch.py b/vcr/patch.py
index 965ef85..60e832b 100644
--- a/vcr/patch.py
+++ b/vcr/patch.py
@@ -212,8 +212,8 @@ class CassettePatcherBuilder:
         get_conn = connection_pool_class._get_conn

         @functools.wraps(get_conn)
-        def patched_get_conn(pool, timeout=None):
-            connection = get_conn(pool, timeout)
+        def patched_get_conn(pool, timeout=None, **kwargs):
+            connection = get_conn(pool, timeout, **kwargs)
             connection_class = (
                 pool.ConnectionCls if hasattr(pool, "ConnectionCls") else connection_class_getter()
             )
@@ -224,7 +224,7 @@ class CassettePatcherBuilder:
             # class) around. This while loop will terminate because
             # eventually the pool will run out of connections.
             while not isinstance(connection, connection_class):
-                connection = get_conn(pool, timeout)
+                connection = get_conn(pool, timeout, **kwargs)
             return connection

         return patched_get_conn
@@ -233,8 +233,8 @@ class CassettePatcherBuilder:
         new_conn = connection_pool_class._new_conn

         @functools.wraps(new_conn)
-        def patched_new_conn(pool):
-            new_connection = new_conn(pool)
+        def patched_new_conn(pool, **kwargs):
+            new_connection = new_conn(pool, **kwargs)
             connection_remover.add_connection_to_pool_entry(pool, new_connection)
             return new_connection

diff --git a/vcr/stubs/__init__.py b/vcr/stubs/__init__.py
index 2c3b8b5..b41ae2f 100644
--- a/vcr/stubs/__init__.py
+++ b/vcr/stubs/__init__.py
@@ -401,6 +407,7 @@ class VCRHTTPConnection(VCRConnection):
     _protocol = "http"
     debuglevel = _baseclass.debuglevel
     _http_vsn = _baseclass._http_vsn
+    _http_vsn_str = getattr(_baseclass, '_http_vsn_str', None)


 class VCRHTTPSConnection(VCRConnection):
@@ -411,3 +418,4 @@ class VCRHTTPSConnection(VCRConnection):
     is_verified = True
     debuglevel = _baseclass.debuglevel
     _http_vsn = _baseclass._http_vsn
+    _http_vsn_str = getattr(_baseclass, '_http_vsn_str', None)
```

What I am stuck on, and why I believe this is the right place to report this, as you're responsible for getting it to work if the idea is 'niquests/urllib3-future are drop in replacements':

```
.tox/py314/lib64/python3.14/site-packages/niquests/sessions.py:1374: in send
    r = adapter.send(request, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.tox/py314/lib64/python3.14/site-packages/niquests/adapters.py:884: in send
    resp_or_promise = conn.urlopen(  # type: ignore[call-overload,misc]
.tox/py314/lib64/python3.14/site-packages/urllib3/connectionpool.py:1897: in urlopen
    assert isinstance(response, HTTPResponse)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E   AssertionError
```

vcrpy subclasses `http.client.HTTPResponse` but urllib3-future's HTTPResponse suclasses io.IOBase. The old urllib3 subclasses http.client.HTTPResponse, thus the conundrum in that assertion.

I would like to see this fixed somehow - compatibility with vcrpy/pytest-recording is the main impediment for me using niquests.

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,3 +1,9 @@
+2.19.908 (2026-04-14)
+=====================
+
+- Fixed a rare racing condition preventing the synchronous happy eyeballs algorithm to complete if all IP addresses are unreachable.
+- Fixed SSE unicode decode error on partial bytes chunk. (#349)
+
 2.19.907 (2026-04-13)
 =====================
 

--- a/src/urllib3/_version.py
+++ b/src/urllib3/_version.py
@@ -1,4 +1,4 @@
 # This file is protected via CODEOWNERS
 from __future__ import annotations
 
-__version__ = "2.19.907"
+__version__ = "2.19.908"

--- a/src/urllib3/backend/_base.py
+++ b/src/urllib3/backend/_base.py
@@ -549,13 +549,15 @@ class BaseBackend:
     @property
     def _http_vsn_str(self) -> str:
         """Reimplemented for backward compatibility purposes."""
-        assert self._svn is not None
+        if self._svn is None:
+            return "HTTP/1.1"
         return self._svn.value
 
     @property
     def _http_vsn(self) -> int:
         """Reimplemented for backward compatibility purposes."""
-        assert self._svn is not None
+        if self._svn is None:
+            return 11
         if self._cached_http_vsn is None:
             self._cached_http_vsn = int(self._svn.value.split("/")[-1].replace(".", ""))
         return self._cached_http_vsn

--- a/src/urllib3/connectionpool.py
+++ b/src/urllib3/connectionpool.py
@@ -593,20 +593,22 @@ class HTTPConnectionPool(ConnectionPool, RequestMethods):
                     event = threading.Event()
                     winning_task: Future[None] | None = None
                     completed_count: int = 0
+                    _heb_callback_lock = threading.Lock()
 
                     def _happy_eyeballs_completed(t: Future[None]) -> None:
                         nonlocal winning_task, event, completed_count
 
-                        if winning_task is None and t.exception() is None:
-                            winning_task = t
-                            event.set()
+                        with _heb_callback_lock:
+                            if winning_task is None and t.exception() is None:
+                                winning_task = t
+                                event.set()
 
-                            return
+                                return
 
-                        completed_count += 1
+                            completed_count += 1
 
-                        if completed_count >= len(challengers):
-                            event.set()
+                            if completed_count >= len(challengers):
+                                event.set()
 
                     tpe = ThreadPoolExecutor(max_workers=max_task)
 
@@ -618,16 +620,29 @@ class HTTPConnectionPool(ConnectionPool, RequestMethods):
 
                         tasks.append(task)
 
-                    event.wait()
+                    event.wait(timeout=override_timeout + 1.0)
 
                     for task in tasks:
                         if task == winning_task:
                             continue
 
+                        associated_conn = challengers[tasks.index(task)]
+
                         if task.running():
+                            try:
+                                if associated_conn._resolver._sock_cursor:
+                                    associated_conn._resolver._sock_cursor.shutdown(0)
+                                    associated_conn._resolver._sock_cursor.close()
+                                elif associated_conn.sock:
+                                    associated_conn.sock.shutdown(0)
+                                    associated_conn.sock.close()
+                                    associated_conn.sock = None
+                            except OSError:
+                                pass
+                            associated_conn.close()
                             task.cancel()
                         else:
-                            challengers[tasks.index(task)].close()
+                            associated_conn.close()
 
                     if winning_task is None:
                         within_delay_msg: str = (
@@ -2235,19 +2250,21 @@ class HTTPSConnectionPool(HTTPConnectionPool):
                     event = threading.Event()
                     winning_task: Future[None] | None = None
                     completed_count: int = 0
+                    _heb_callback_lock = threading.Lock()
 
                     def _happy_eyeballs_completed(t: Future[None]) -> None:
                         nonlocal winning_task, event, completed_count
 
-                        if winning_task is None and t.exception() is None:
-                            winning_task = t
-                            event.set()
-                            return
+                        with _heb_callback_lock:
+                            if winning_task is None and t.exception() is None:
+                                winning_task = t
+                                event.set()
+                                return
 
-                        completed_count += 1
+                            completed_count += 1
 
-                        if completed_count >= len(challengers):
-                            event.set()
+                            if completed_count >= len(challengers):
+                                event.set()
 
                     tpe = ThreadPoolExecutor(max_workers=max_task)
 
@@ -2259,7 +2276,7 @@ class HTTPSConnectionPool(HTTPConnectionPool):
 
                         tasks.append(task)
 
-                    event.wait()
+                    event.wait(timeout=override_timeout + 1.0)
 
                     for task in tasks:
                         if task == winning_task:

--- a/src/urllib3/connectionpool.py
+++ b/src/urllib3/connectionpool.py
@@ -727,6 +727,8 @@ class HTTPConnectionPool(ConnectionPool, RequestMethods):
             with self.pool._lock:
                 if self.pool.busy_with_placeholder:
                     self.pool.kill_cursor()
+                elif self.pool.busy:
+                    self.pool.release()
                 self.pool.put(conn, immediately_unavailable=True)
             return conn
 
@@ -1741,7 +1743,11 @@ class HTTPConnectionPool(ConnectionPool, RequestMethods):
         try:
             # Request a connection from the queue.
             timeout_obj = self._get_timeout(timeout)
-            conn = self._get_conn(timeout=pool_timeout, heb_timeout=timeout_obj)
+
+            try:
+                conn = self._get_conn(timeout=pool_timeout, heb_timeout=timeout_obj)
+            except TypeError:  # Defensive: 3rd party conn hacks
+                conn = self._get_conn(timeout=pool_timeout)
 
             conn.timeout = timeout_obj.connect_timeout  # type: ignore[assignment]
 

--- a/src/urllib3/contrib/webextensions/_async/sse.py
+++ b/src/urllib3/contrib/webextensions/_async/sse.py
@@ -1,5 +1,6 @@
 from __future__ import annotations
 
+import codecs
 import typing
 
 if typing.TYPE_CHECKING:
@@ -16,6 +17,7 @@ class AsyncServerSideEventExtensionFromHTTP(AsyncExtensionFromHTTP):
 
         self._last_event_id: str | None = None
         self._buffer: str = ""
+        self._decoder = codecs.getincrementaldecoder("utf-8")()
         self._stream: typing.AsyncGenerator[bytes, None] | None = None
 
     @staticmethod
@@ -74,11 +76,15 @@ class AsyncServerSideEventExtensionFromHTTP(AsyncExtensionFromHTTP):
             # (terminated by a blank line: \n\n or \r\n\r\n).
             while "\n\n" not in self._buffer and "\r\n\r\n" not in self._buffer:
                 try:
-                    self._buffer += (await self._stream.__anext__()).decode("utf-8")
+                    self._buffer += self._decoder.decode(await self._stream.__anext__())
                 except StopAsyncIteration:
-                    await self._stream.aclose()
-                    self._stream = None
-                    return None
+                    last_chunk = self._decoder.decode(b"", final=True)
+                    if not last_chunk:
+                        await self._stream.aclose()
+                        self._stream = None
+                        self._decoder.reset()
+                        return None
+                    self._buffer += last_chunk
 
             # Locate the first event boundary.
             lf = self._buffer.find("\n\n")

--- a/src/urllib3/contrib/webextensions/sse.py
+++ b/src/urllib3/contrib/webextensions/sse.py
@@ -1,5 +1,6 @@
 from __future__ import annotations
 
+import codecs
 import json
 import typing
 from threading import RLock
@@ -69,6 +70,7 @@ class ServerSideEventExtensionFromHTTP(ExtensionFromHTTP):
 
         self._last_event_id: str | None = None
         self._buffer: str = ""
+        self._decoder = codecs.getincrementaldecoder("utf-8")()
         self._lock = RLock()
         self._stream: typing.Generator[bytes, None, None] | None = None
 
@@ -128,10 +130,14 @@ class ServerSideEventExtensionFromHTTP(ExtensionFromHTTP):
                 # event (terminated by a blank line: \n\n or \r\n\r\n).
                 while "\n\n" not in self._buffer and "\r\n\r\n" not in self._buffer:
                     try:
-                        self._buffer += next(self._stream).decode("utf-8")
+                        self._buffer += self._decoder.decode(next(self._stream))
                     except StopIteration:
-                        self._stream = None
-                        return None
+                        last_chunk = self._decoder.decode(b"", final=True)
+                        if not last_chunk:
+                            self._stream = None
+                            self._decoder.reset()
+                            return None
+                        self._buffer += last_chunk
 
                 # Locate the first event boundary.
                 lf = self._buffer.find("\n\n")

--- a/src/urllib3/util/_async/traffic_police.py
+++ b/src/urllib3/util/_async/traffic_police.py
@@ -1187,6 +1187,9 @@ class AsyncTrafficPolice(typing.Generic[T]):
     def rsize(self) -> int:
         return len(self._registry)
 
+    def empty(self) -> bool:
+        return bool(self.rsize())
+
     async def beacon(self, traffic_indicator: MappableTraffic | type) -> bool:
         if not isinstance(traffic_indicator, type):
             key: PoolKey | int = (

--- a/src/urllib3/util/traffic_police.py
+++ b/src/urllib3/util/traffic_police.py
@@ -999,6 +999,9 @@ class TrafficPolice(typing.Generic[T]):
         with self._lock:
             return len(self._registry)
 
+    def empty(self) -> bool:
+        return bool(self.rsize())
+
     def beacon(self, traffic_indicator: MappableTraffic | type) -> bool:
         """Answer the following question: Do this PoliceTraffic know about the traffic_indicator?"""
         if not isinstance(traffic_indicator, type):

--- a/test/with_traefik/test_thread_safety.py
+++ b/test/with_traefik/test_thread_safety.py
@@ -71,7 +71,9 @@ class TestThreadSafety(TraefikTestCase):
             responses = []
             for _ in range(16):
                 try:
-                    responses.append(s.urlopen("GET", f"{self.https_url}/get"))
+                    responses.append(
+                        s.urlopen("GET", f"{self.https_url}/get", timeout=10.0)
+                    )
                 except Exception as e:
                     print(e)
                     assert False

